Description: Use the renamed yasm in Lucid
Author: Chris Coulson <chris.coulson@canonical.com>
Forwarded: not-needed

Index: mozilla/configure.in
===================================================================
--- mozilla.orig/configure.in	2012-02-21 02:03:59.000000000 +0000
+++ mozilla/configure.in	2012-02-21 21:23:28.896760911 +0000
@@ -4403,7 +4403,7 @@
 dnl ======================
 
 AC_MSG_CHECKING([for YASM assembler])
-AC_CHECK_PROGS(YASM, yasm, "")
+AC_CHECK_PROGS(YASM, yasm-1, "")
 
 if test -n "$YASM"; then
   dnl Pull out yasm's version string
@@ -4411,7 +4411,7 @@
   _YASM_VER_FILTER='s|.* \([0-9]\{1,\}\.[0-9]\{1,\}\.[0-9]\{1,\}\.[0-9]\{1,\}\).*|\1|p'
   changequote([,])
 
-  YASM_VERSION=`yasm --version | sed -ne "$_YASM_VER_FILTER"`
+  YASM_VERSION=`yasm-1 --version | sed -ne "$_YASM_VER_FILTER"`
   _YASM_MAJOR_VERSION=`echo ${YASM_VERSION} | $AWK -F\. '{ print $1 }'`
   _YASM_MINOR_VERSION=`echo ${YASM_VERSION} | $AWK -F\. '{ print $2 }'`
   _YASM_RELEASE=`      echo ${YASM_VERSION} | $AWK -F\. '{ print $3 }'`
@@ -6235,7 +6235,7 @@
 dnl it doesn't exist or we have too old of a version.
 if test -n "$LIBJPEG_TURBO_X86_ASM" -o -n "$LIBJPEG_TURBO_X64_ASM" ; then
     AC_MSG_CHECKING([for Yasm assembler])
-    AC_CHECK_PROGS(LIBJPEG_TURBO_AS, yasm, "")
+    AC_CHECK_PROGS(LIBJPEG_TURBO_AS, yasm-1, "")
 
     if test -z "$LIBJPEG_TURBO_AS" ; then
         AC_MSG_ERROR([Yasm is required to build with libjpeg-turbo's optimized JPEG decoding routines, but you do not appear to have Yasm installed.  Either install it or configure with --disable-libjpeg-turbo to use the pure C JPEG decoder.  See https://developer.mozilla.org/en/YASM for more details.])
Index: mozilla/config/config.mk
===================================================================
--- mozilla.orig/config/config.mk	2012-02-21 02:03:58.000000000 +0000
+++ mozilla/config/config.mk	2012-02-21 21:23:28.896760911 +0000
@@ -176,7 +176,7 @@
 endif
 
 ifneq (,$(MOZ_DEBUG)$(MOZ_DEBUG_SYMBOLS))
-  ifeq ($(AS),yasm)
+  ifeq ($(AS),yasm-1)
     ifeq ($(OS_ARCH)_$(GNU_CC),WINNT_)
       _DEBUG_ASFLAGS += -g cv8
     else
Index: mozilla/js/src/config/config.mk
===================================================================
--- mozilla.orig/js/src/config/config.mk	2012-02-21 21:26:52.224757298 +0000
+++ mozilla/js/src/config/config.mk	2012-02-21 21:27:13.112756927 +0000
@@ -176,7 +176,7 @@
 endif
 
 ifneq (,$(MOZ_DEBUG)$(MOZ_DEBUG_SYMBOLS))
-  ifeq ($(AS),yasm)
+  ifeq ($(AS),yasm-1)
     ifeq ($(OS_ARCH)_$(GNU_CC),WINNT_)
       _DEBUG_ASFLAGS += -g cv8
     else
